packages/core/src/types.ts

```diff
@@ -423,6 +424,26 @@ export function getElementDefinition(typeName: string, propertyName: string): El
   return property;
 }
 
+/**
+ * Typeguard to validate that an object is a FHIR resource
+ * @param value The object to check
+ * @returns True if the input is of type 'object' and contains property 'resourceType'
+ */
+export function isResource<T extends Resource = Resource>(value: T | Reference<T> | string | undefined): value is T {
```

Comment:
We may want to consider adding `null` to the function signature, in which case we would also need to add a null check in the function body because `typeof null === 'object'` 😡

You could go even further and change the function signature to `value: unknown`, which would accept anything.  That would be nice, because then you could throw anything at it, and it will tell you if it's a resource or not.  That might break the generic `<T>` return type though.